prometheus_client/registry.py

```diff
@@ -135,10 +135,20 @@ def __init__(self, names, registry):
         self._registry = registry
 
     def collect(self):
-        for metric in self._registry.collect():
-            m = metric._restricted_metric(self._name_set)
-            if m:
-                yield m
+        names = copy.copy(self._name_set)
+        collectors = set()
+        with self._registry._lock:
+            if 'target_info' in names and self._registry._target_info:
+                yield self._registry._target_info_metric()
```

Comment:
I don't think it is safe to yield under the lock.

Reply:
Fixed. Good catch!